typing/typecore.ml

```diff
@@ -2541,40 +2541,40 @@ let generalize_and_check_univars env kind exp ty_expected vars =
    performed, then the check(s) are delayed until after typechecking in order to
    give a chance for the type to become known through unification. *)
 
+let check_statement exp =
+  let ty = get_desc (expand_head exp.exp_env exp.exp_type) in
+  match ty with
+  | Tconstr (p, _, _)  when Path.same p Predef.path_unit -> ()
+  | Tvar _ -> ()
+  | _ ->
+      let rec loop {exp_loc; exp_desc; exp_extra; _} =
+        match exp_desc with
+        | Texp_let (_, _, e)
+        | Texp_sequence (_, e)
+        | Texp_letexception (_, e)
+        | Texp_letmodule (_, _, _, _, e) ->
+            loop e
+        | _ ->
+            let loc =
+              match List.find_opt (function
+                  | (Texp_constraint _, _, _) -> true
+                  | _ -> false) exp_extra
+              with
+              | Some (_, loc, _) -> loc
+              | None -> exp_loc
+            in
+            Location.prerr_warning loc Warnings.Non_unit_statement
+      in
+      loop exp
+
 let check_partial_application statement exp =
   let ty = get_desc (expand_head exp.exp_env exp.exp_type) in
-  let check_statement () =
-    match ty with
-    | Tconstr (p, _, _)  when Path.same p Predef.path_unit -> ()
-    | Tvar _ -> ()
-    | _ ->
-        if statement then
-          let rec loop {exp_loc; exp_desc; exp_extra; _} =
-            match exp_desc with
-            | Texp_let (_, _, e)
-            | Texp_sequence (_, e)
-            | Texp_letexception (_, e)
-            | Texp_letmodule (_, _, _, _, e) ->
-                loop e
-            | _ ->
-                let loc =
-                  match List.find_opt (function
-                      | (Texp_constraint _, _, _) -> true
-                      | _ -> false) exp_extra
-                  with
-                  | Some (_, loc, _) -> loc
-                  | None -> exp_loc
-                in
-                Location.prerr_warning loc Warnings.Non_unit_statement
-          in
-          loop exp
-  in
   match ty with
   | Tarrow _ ->
       let rec check {exp_desc; exp_loc; exp_extra; _} =
         if List.exists (function
             | (Texp_constraint _, _, _) -> true
-            | _ -> false) exp_extra then check_statement ()
+            | _ -> false) exp_extra && statement then check_statement exp
```

Comment:
(Looking at the 'lift check_statement out of ..') I think the logic changes here: if `statement` is false, the new code goes to the `else` branch, while before `check_statement ()` would do nothing. (This is the difference between `if foo then (if statement then bar) else baz` and `if foo && statement then bar else baz`.)

Reply:
Oops you are right here. However, the commits that follow "fix" the bug introduced here...